xdesign/acquisition.py

```diff
@@ -540,3 +549,183 @@ def raster_scan(sx, sy):
+        dset[:, :, :] = np.zeros(dset.shape)
+        f.close()
+
+
+class Simulator(object):
+    """Optical simulation based on multislice propagation.
+    
+    Attributes
+    ----------
+    grid : numpy.ndarray or list of numpy.ndarray
+        Descretized grid for the phantom object. If type == 'refractive_index', 
+        it takes a list of [delta_grid, beta_grid].
+    energy : float
+        Beam energy in keV. Should match the energy used for creating the grids.
+    psize : list
+        Pixel size in nm.
+    type : str
+        Value type of input grid. 
+    """
+
+    def __init__(self, energy, grid=None, psize=None, type='refractive_index'):
+
+        if type == 'refractive_index':
+            if grid is not None:
+                self.grid_delta, self.grid_beta = grid
+            else:
+                self.grid_delta = self.grid_beta = None
+            self.energy = energy
+            self.voxel = psize
+            self._ndim = self.grid_delta.ndim
+            self.size = self.grid_delta.shape
+            self.lmbda_nm = 1.24 / self.energy
+            self.mesh = []
+            temp = []
+            for i in range(self._ndim):
+                temp.append(np.arange(self.size[i]))
+            self.mesh = np.meshgrid(*temp, indexing='ij')
+
+            # wavefront in x-y plane or x edge
+            self.wavefront = np.zeros(self.grid_delta.shape[:-1], dtype=np.complex64)
+        else:
+            raise ValueError('Currently only delta and beta grids are supported.')
+
+    def save_grid(self, save_path='data/sav/grid'):
+        if not os.path.exists(save_path):
+            os.makedirs(save_path)
+        np.save(os.path.join(save_path, 'grid_delta'), self.grid_delta)
+        np.save(os.path.join(save_path, 'grid_beta'), self.grid_beta)
+        grid_pars = [self.size, self.voxel, self.energy]
+        np.save(os.path.join(save_path, 'grid_pars'), grid_pars)
+
+    def read_grid(self, save_path='data/sav/grid'):
+        try:
+            self.grid_delta = np.load(os.path.join(save_path, 'grid_delta.npy'))
+            self.grid_beta = np.load(os.path.join(save_path, 'grid_beta.npy'))
+        except:
+            raise ValueError('Failed to read grid.')
+
+    def save_slice_images(self, save_path='data/sav/slices'):
+        if not os.path.exists(save_path):
+            os.makedirs(save_path)
+        dxchange.write_tiff_stack(self.grid_delta, os.path.join(save_path, 'delta'),
+                                  overwrite=True, dtype=np.float32)
+        dxchange.write_tiff_stack(self.grid_beta, os.path.join(save_path, 'beta'),
+                                  overwrite=True, dtype=np.float32)
+
+    def show_grid(self, part='delta'):
+        if part == 'delta':
+            tifffile.imshow(self.grid_delta)
+        elif part == 'beta':
+            tifffile.imshow(self.grid_beta)
+        else:
+            print('WARNING: wrong part specified for show_grid.')
```

Comment:
Use the [`warnings`](https://docs.python.org/3/library/warnings.html) module for warnings.